contracts/FractalUsul.sol

```diff
@@ -1,86 +1,469 @@
-  /// @param strategy Address of Voting Strategy, under which proposal submitted
-  /// @param data - any additional data, which would be passed into IStrategy.receiveProposal
-  /// @param transactions - array of transactions to execute
-  /// @param title - proposal title, emitted in ProposalMetadataCreated
-  /// @param description - proposal description, emitted in ProposalMetadataCreated
-  /// @param documentationUrl - proposal documentation/discussion URL, emitted in ProposalMetadataCreated. 
-  /// Supposed to be link to Discord/Slack/Whatever chat discussion
-  function submitProposalWithMetaData(
+import "@gnosis.pm/zodiac/contracts/core/Module.sol";
+import "./usul/IBaseStrategy.sol";
+import "./interfaces/IFractalUsul.sol";
+
+/// @title FractalUsul - A Zodiac module that enables a voting agnostic proposal mechanism
+contract FractalUsul is Module, IFractalUsul {
+    struct Transaction {
+        address to;
+        uint256 value;
+        bytes data;
+        Enum.Operation operation;
+    }
+
+    struct Proposal {
+        bool canceled;
+        uint256 timelockPeriod;
+        bytes32[] txHashes;
+        uint256 executionCounter;
+        address strategy; // the module that is allowed to vote on this
+    }
+
+    bytes32 public constant DOMAIN_SEPARATOR_TYPEHASH =
+        0x47e79534a245952e8b16893a336b85a3d9ea9fa8c573f3d803afb92a79469218;
+    bytes32 public constant TRANSACTION_TYPEHASH =
+        0x72e9670a7ee00f5fbf1049b8c38e3f22fab7e9b85029e85cf9412f17fdd5c2ad;
+    uint256 public totalProposalCount; // total number of submitted proposals
+    address internal constant SENTINEL_STRATEGY = address(0x1);
+    mapping(uint256 => Proposal) internal proposals; // Proposals by proposal ID
+    mapping(address => address) internal strategies;
+
+    event ProposalCreated(
+        address strategy,
+        uint256 proposalId,
+        address proposer,
+        Transaction[] transactions,
+        string title,
+        string description,
+        string documentationUrl
+    );
+    event ProposalCanceled(uint256 proposalId);
+    event TransactionExecuted(uint256 proposalId, bytes32 txHash);
+    event TransactionExecutedBatch(uint256 startIndex, uint256 endIndex);
+    event ProposalTimelocked(uint256 proposalId, uint256 endDate);
+    event ProposalExecuted(uint256 id);
+    event UsulSetup(
+        address indexed initiator,
+        address indexed owner,
+        address indexed avatar,
+        address target
+    );
+    event EnabledStrategy(address strategy);
+    event DisabledStrategy(address strategy);
+
+    constructor(
+        address _owner,
+        address _avatar,
+        address _target,
+        address[] memory _strategies
+    ) {
+        bytes memory initParams = abi.encode(
+            _owner,
+            _avatar,
+            _target,
+            _strategies
+        );
+        setUp(initParams);
+    }
+
+    function setUp(bytes memory initParams) public override initializer {
```

Comment:
What does the initializer modifier do? Is that only callable once?

Reply:
Correct, it makes the function only callable once

Reply:
cool cool